dropshot/src/error.rs

```diff
@@ -316,6 +343,657 @@ impl Error for HttpError {
+        INTERNAL_SERVER_ERROR;
+        /// 501 Not Implemented
+        /// [[RFC7231, Section 6.6.2](https://tools.ietf.org/html/rfc7231#section-6.6.2)]
+        NOT_IMPLEMENTED;
+        /// 502 Bad Gateway
+        /// [[RFC7231, Section 6.6.3](https://tools.ietf.org/html/rfc7231#section-6.6.3)]
+        BAD_GATEWAY;
+        /// 503 Service Unavailable
+        /// [[RFC7231, Section 6.6.4](https://tools.ietf.org/html/rfc7231#section-6.6.4)]
+        SERVICE_UNAVAILABLE;
+        /// 504 Gateway Timeout
+        /// [[RFC7231, Section 6.6.5](https://tools.ietf.org/html/rfc7231#section-6.6.5)]
+        GATEWAY_TIMEOUT;
+        /// 505 HTTP Version Not Supported
+        /// [[RFC7231, Section 6.6.6](https://tools.ietf.org/html/rfc7231#section-6.6.6)]
+        HTTP_VERSION_NOT_SUPPORTED;
+        /// 506 Variant Also Negotiates
+        /// [[RFC2295](https://tools.ietf.org/html/rfc2295)]
+        VARIANT_ALSO_NEGOTIATES;
+        /// 507 Insufficient Storage
+        /// [[RFC4918](https://tools.ietf.org/html/rfc4918)]
+        INSUFFICIENT_STORAGE;
+        /// 508 Loop Detected
+        /// [[RFC5842](https://tools.ietf.org/html/rfc5842)]
+        LOOP_DETECTED;
+
+        /// 510 Not Extended
+        /// [[RFC2774](https://tools.ietf.org/html/rfc2774)]
+        NOT_EXTENDED;
+        /// 511 Network Authentication Required
+        /// [[RFC6585](https://tools.ietf.org/html/rfc6585)]
+        NETWORK_AUTHENTICATION_REQUIRED;
+    }
+
+    /// Converts an [`http::StatusCode`] into an [`ErrorStatusCode`].
+    ///
+    /// # Returns
+    ///
+    /// - [`Ok`]`(`[`ErrorStatusCode`]`)` if the status code is a 4xx or 5xx
+    ///   status code.
+    /// - [`Err`]`(`[`NotAnError`]`)` if the status code is a 1xx, 2xx, or 3xx
+    ///   status code.
+    pub fn from_status(status: http::StatusCode) -> Result<Self, NotAnError> {
+        if status.is_client_error() || status.is_server_error() {
+            Ok(ErrorStatusCode(status))
+        } else {
+            Err(NotAnError(status))
+        }
+    }
+
+    /// Converts a u16 to a status code.
+    ///
+    /// The function validates the correctness of the supplied `u16` It must
+    /// be a HTTP client error (400-499) or server error (500-599).
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use dropshot::ErrorStatusCode;
+    ///
+    /// // 404 is a client error
+    /// let ok = ErrorStatusCode::from_u16(404).unwrap();
+    /// assert_eq!(ok, ErrorStatusCode::NOT_FOUND);
+    ///
+    /// // 555 is a server error (although it lacks a well known meaning)
+    /// let _ok = ErrorStatusCode::from_u16(555).unwrap();
+    ///
+    /// // 200 is a status code, but not an error.
+    /// let err = ErrorStatusCode::from_u16(200);
+    /// assert!(err.is_err());
+    ///
+    /// // 99 is out of range for any status code
+    /// let err = ErrorStatusCode::from_u16(99);
+    /// assert!(err.is_err());
+    /// ```
+    #[inline]
```

Comment:
The `#[inline]` attributes are here just because I lifted the function signatures for these conversions straight from the `http` crate and then modified them, it inlines these conversions. I'm happy to remove them, I doubt they have much of an impact.

Reply:
Aha! That makes a ton of sense and I'm happy to leave them (or keep them). Whatever you prefer.